receiver/routereceiver/receiver.go

```diff
@@ -0,0 +1,180 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routereceiver
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+var (
+	// routes is a map of registered routes.
+	routes = map[string]*receiver{}
+
+	// mux is a mutex for accessing receivers.
+	mux sync.RWMutex
+
+	// errMetricPipelineNotDefined is returned when a metric pipeline is not set.
+	errMetricPipelineNotDefined = errors.New("metric pipeline not defined for route")
+
+	// errLogPipelineNotDefined is returned when a log pipeline is not set.
+	errLogPipelineNotDefined = errors.New("log pipeline not defined for route")
+
+	// errTracePipelineNotDefined is returned when a trace pipeline is not set.
+	errTracePipelineNotDefined = errors.New("trace pipeline not defined for route")
+
+	// errReceiverNotSet is an error returned when a receiver is not set.
+	errRouteNotDefined = errors.New("route not defined")
+)
+
+// receiver is a struct that receives routed telemetry.
+type receiver struct {
+	name           string
+	stopOnce       sync.Once
+	metricConsumer consumer.Metrics
+	logConsumer    consumer.Logs
+	traceConsumer  consumer.Traces
+}
+
+// Start starts the receiver.
+func (r *receiver) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+// Shutdown stops the receiver.
+func (r *receiver) Shutdown(_ context.Context) error {
+	r.stopOnce.Do(func() {
```

Comment:
I don't think you need this in a stopOnce. `removeRoute` wraps [delete](https://pkg.go.dev/builtin#delete) which is a noop if there is nothing to delete. I'm not sure if this could be called up to three times if it's used in logs, traces, and metrics but even so two of those calls would be noops.

Reply:
Sounds good. I'll remove this then. I was following this pattern from the OTLP receiver which appears to do this to avoid unregistering the receiver 3 times. But I agree. It's unnecessary.